### ISSUE TITLE
Cilium: can't update network policy, proxy disabled

### ISSUE THREAD

**Issue by borkmann** [cilium/cilium#9100]
Seen on master, container cannot be created:
```
[...]
level=debug msg="Error regenerating endpoint: can't update network policy, proxy disabled" code=Failure containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=683 endpointState=ready identity=5 ipv4=192.168.195.205 ipv6="2001:db8:5::1813" k8sPodName=/ policyRevision=0 subsys=endpoint type=200
level=debug msg="EventQueue event processing statistics" eventConsumeOffQueueWaitTime="53.842µs" eventEnqueueWaitTime=908ns eventHandlingDuration=1.673356ms eventType="*endpoint.EndpointRegenerationEvent" name=endpoint-683 subsys=eventqueue
level=debug msg="Controller func execution time: 1.739864ms" name="sync-IPv6-identity-mapping (683)" subsys=controller uuid=82b0a897-cd64-11e9-aa95-0cc47a8629fc
level=error msg="endpoint regeneration failed" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=683 error="can't update network policy, proxy disabled" identity=5 ipv4=192.168.195.205 ipv6="2001:db8:5::1813" k8sPodName=/ subsys=endpoint
level=debug msg="Controller func execution time: 2.126695ms" name="sync-IPv4-identity-mapping (683)" subsys=controller uuid=82b0a799-cd64-11e9-aa95-0cc47a8629fc
level=debug msg="Received event" key="cilium/state/ip/v1/default/2001:db8:5::1813" kvstore-event=create subsys=ipcache
[...]
```
Agent is run as:
```
./daemon/cilium-agent --ipv4-range 192.168.0.0/16 --ipv6-range 2001:db8:5::/96 --kvstore-opt consul.address=147.75.101.11:8500 --kvstore consul --enable-ipv4=true --enable-ipv6=true --disable-envoy-version-check=true --masquerade=false --ipv4-node=147.75.101.11 --install-iptables-rules=false -D
```
There's no policy installed.
```
# docker run --name foo --network cilium-net -td httpd
8cae6bf1ccde96864b59b4f98710cad004a849911dbdc424347cb399658c3132
docker: Error response from daemon: failed to create endpoint foo7 on network cilium-net: NetworkDriver.CreateEndpoint: Error creating endpoint Put http:///var/run/cilium/cilium.sock/v1/endpoint/cilium-local:0: context deadline exceeded.
```
Expectation would be that regeneration does not fail.

/cc @jrajahalme

**Comment by borkmann** [comment]
Workaround I have locally for the regeneration:
```
diff --git a/daemon/daemon.go b/daemon/daemon.go
index d4008cf..7792c9b 100644
--- a/daemon/daemon.go
+++ b/daemon/daemon.go
@@ -227,7 +227,7 @@ func (d *Daemon) RemoveProxyRedirect(e regeneration.EndpointInfoSource, id strin
 func (d *Daemon) UpdateNetworkPolicy(e regeneration.EndpointUpdater, policy *policy.L4Policy,
        proxyWaitGroup *completion.WaitGroup) (error, revert.RevertFunc) {
        if d.l7Proxy == nil {
-               return fmt.Errorf("can't update network policy, proxy disabled"), nil
+               return nil, nil
        }
        err, revertFunc := d.l7Proxy.UpdateNetworkPolicy(e, policy, e.GetIngressPolicyEnabledLocked(),
                e.GetEgressPolicyEnabledLocked(), proxyWaitGroup)
```

**Comment by jrajahalme** [comment]
@borkmann I think the disabling of the error return should be gated by `if !policy.HasRedirect()` so that an error would still be returned if there is a policy that requires a redirect.

**Comment by brb** [comment]
A user just reported that they hit the same issue with the following configuration:

```
helm template cilium \
  --namespace kube-system \
  --set global.containerRuntime.integration=crio \
  --set global.datapathMode=ipvlan \
  --set global.ipvlan.masterDevice=eth0 \
  --set global.tunnel=disabled \
  --set global.masquerade=true \
  --set global.installIptablesRules=false \
  --set global.autoDirectNodeRoutes=true \
  > cilium.yaml
```

https://cilium.slack.com/archives/C1MATJ5U5/p1568740546033500?thread_ts=1568738884.029500&cid=C1MATJ5U5

**Comment by aanm** [comment]
to whomever will fix this. PTAL at https://github.com/cilium/cilium/pull/9157/files#r322510437

**Comment by tgraf** [comment]
Fixed by #9481
